Update Johnson's algorithm so that graph is not modified(#1506)

theosotr opened the pull request:
In response to https://github.com/networkx/networkx/issues/1506 this pull request updates ``johnson`` in order graph not be modified.

To be more specific, not any node is added to the graph to compute dist values of each other node using Bellman - Ford algorithm. For this purpose, a new function is added ``_bellman_ford_relaxation`` which contains the relaxation loop for Bellman - Ford. It is used both in ``johnson`` and ``bellman_ford``

theosotr commented:
@ysitu What is your opinion?

ysitu commented:
Will take a look later today.

ysitu commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -584,6 +584,23 @@ def bellman_ford(G, source, weight='weight'):
+
+def _bellman_ford_relaxation(G, pred, dist, source_list, weight):
+    """
+    Relaxation loop for Bellman - Ford algorithm.
```
Punctuation: `Bellman-Ford`.

jfinkels replied:
Actually, it should be an en-dash: Bellman–Ford algorithm

ysitu commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -584,6 +584,23 @@ def bellman_ford(G, source, weight='weight'):
+    :param weight: Edge data key corresponding to the edge weight
+    :return: dictionaries
+       Returns two dictionaries keyed by node to predecessor in the
+       path and to the distance from the source respectively.
```
Please follow the examples of other functions in formatting the docstring.

theosotr replied:
This function is supposed to be auxiliary. This is the reason why I did not  follow the official way of formatting the docstring.

ysitu commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -584,6 +584,23 @@ def bellman_ford(G, source, weight='weight'):
+    :param pred: dictionary keyed by node to predecessor in the
+    path
+    :param dist: dictionary keyed by node to distance from the source
+    :param source_list: list of nodes setting as source nodes
```
"Iterable of source nodes"

ysitu commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -1044,21 +1060,22 @@ def johnson(G, weight='weight', new_weight=None):
+        new_weight = str(uuid.uuid1())
 
     for u, v, w in G.edges(data=True):
         w[new_weight] = w[weight] + dist[u] - dist[v]
```
By doing this you are still modifying the graph.

theosotr replied:
I believed that the point was not to add a new node to the graph. So, one thought is to keep new weights in a dictionary and then pass it as input of ``_djikstra_relaxation`` function.

ysitu replied:
I think that we should take this opportunity to enable edge weights to be specified using a lambda. Existing code already uses a function internally to handle Graph vs MultiGraph. It should be straightforward to generalize.

theosotr replied:
And passing lambda as parameter. Am I wrong?

ysitu replied:
No, you are correct.

theosotr commented:
I added auxiliary function ``_dijkstra_relaxation`` to remove duplicate code of Dijkstra algorithm. Moreover, I updated ``johnson`` not to modify a given graph. @ysitu What is your opinion?

ysitu commented:
I think that it is better to let `get_weight` have the interface

```python
def get_weight(u, v, data):  # pass data = G[u][v] to save a dict lookup
    ...
```

There is no need for `arg` as that can be folded in `get_weight` by the caller.

`_dijkstra_relaxation` is a misnomer. It is a full implementation of Dijkstra's algorithm. It can be exposed as a public function, too.

theosotr commented:
I defined ``_dijkstra_relaxation`` as an interface for all the different outputs of Dijkstra's algorithm (``dijkstra_predecessor_and_distance``, ``single_source_dijkstra``, etc). Of cource, it can be exposed as a public function, but I wanted to respect the existence of functions mentioned above. However, do you agree with 'philosophy' of the proposed implementation?

theosotr commented:
Αlternatively, ``_dijkstra_relaxation`` can be renamed to ``_dijkstra`` and keep it as 'local' function.

ysitu commented:
I am not asking you to remove all existing functions. Your approach to use one single implementation to support all those functions is a good one. But I feel that it may be useful to also create a solution for #762.

theosotr commented:
> There is no need for ``arg`` as that can be folded in ``get_weight`` by the caller.

As far as this is concerned, I am a little confused how we are going to create generic function ``get_weight``. So far, I thought that this function should be passed as parameter on function which implements Dijkstra's algorithm using lambdas. However, I cannot understand exactly your approach because I am not familliar with all deep and advanced topics in Python. For example, I cannot think a way not to pass dictionary with new edge weights as an argument in case of  Johnson's algorithm. Therefore, could you please explain to me how are you thinking generic function ``get_weight`` should be designed for the calculation of edge weight in different cases (multigraphs, graphs, Johnon's algorithm).

ysitu commented:
If you need to call `get_weight(arg, ...)`, the user can pass in `functools.partial(get_weight, arg)` instead, then you can call `get_weight(...)` without `arg`.

One more thing, I think you can add handling for the case where `get_weight` returns `None`. That can be used to indicate/pretend nonexistence of an edge.

theosotr commented:
I think that creation of a function ``get_weight`` is slower than using lambda because I think that an iteration over edges in order to initialize data dictionary has to be completed.

ysitu commented:
I hope this explains:

```python
def single_source_dijkstra(G, source, target=None, cutoff=None, weight='weight'):
    if source == target:
        return ({source: 0}, {source: [source]})

    if G.is_multigraph():
        get_weight = lambda u, v, edge: min(eattr.get(weight, 1)
                                            for eattr in edge.values())
    else:
        get_weight = lambda u, v, edge: edge.get(weight, 1)

    paths = {source: [source]}  # dictionary of paths
    return _dijkstra_impl(G, source, get_weight, paths=paths,
                          cutoff=cutoff, target=target)


def _dijkstra_impl(G, source, get_weight, pred=None, paths=None, cutoff=None,
                   target=None):
    if G.is_directed():
        G_succ = G.succ
    else:
        G_succ = G.adj

    push = heappush
    pop = heappop
    dist = {}  # dictionary of final distances
    seen = {source: 0}
    c = count()
    fringe = []  # use heapq with (distance,label) tuples
    push(fringe, (0, next(c), source))
    while fringe:
        (d, _, v) = pop(fringe)
        if v in dist:
            continue  # already searched this node.
        dist[v] = d
        if v == target:
            break
        for u, e in G_succ[v].items():
            cost = get_weight(u, v, e)
            if cost is None:
                continue
            vw_dist = dist[v] + cost
            if cutoff is not None:
                if vw_dist > cutoff:
                    continue
            if u in dist:
                if vw_dist < dist[u]:
                    raise ValueError('Contradictory paths found:',
                                     'negative weights?')
            elif u not in seen or vw_dist < seen[u]:
                seen[u] = vw_dist
                push(fringe, (vw_dist, next(c), u))
                if paths is not None:
                    paths[u] = paths[v] + [u]
                if pred is not None:
                    pred[u] = [v]
            elif vw_dist == seen[u]:
                if pred is not None:
                    pred[u].append(v)

    if paths is not None:
        return (dist, paths)
    if pred is not None:
        return (pred, dist)
    return dist
```

dschult commented:
How does:
```python
for u, e in G_succ[v].items():
            cost = get_weight(u, v, e)
```
compare to:
```python
for _, u, cost in G.edges(v, data=weight):```

It's probably better for the long term to avoid exposing the data structure by using G.succ[v].items and friends unless there are performance issues.

ysitu commented:
@dschult Slower performance is the price for abstraction, but it is worth it in the long run. If we have time, we really should rename `G.succ` to `G._succ` (and other dicts) to indicate that it is truly private.

theosotr commented:
My thought was something like this in order to avoid passing ``e`` argument. 

```python
if G.is_multigraph():
        get_weight = lambda u, v: min(eattr.get(weight, 1)
                   for eattr in G.get_edge_data(u, v).values())
else:
        get_weight = lambda u, v: G.get_edge_data(u, v).get(weight, 1)
```

ysitu commented:
Passing the edge data to `get_weight` saves the cost of extra dict lookups in `get_edge_data`.

theosotr commented:
Yes, you are right!

ysitu commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -329,31 +356,35 @@ def single_source_dijkstra(G, source, target=None, cutoff=None, weight='weight')
+            cost = get_weight(v, u, e)
+            if cost is None:
+                continue
+            vw_dist = dist[v] + get_weight(v, u, e)
```
Please rename `vw_dist` to `vu_dist`.

ysitu commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -584,6 +586,41 @@ def bellman_ford(G, source, weight='weight'):
+    dist: dict
+        Keyed by node to the distance from the source
+
+    source_iter: iter
```
For consistency with 2.0 API, please do not remove `_iter` suffix.

ysitu commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -1044,26 +1077,27 @@ def johnson(G, weight='weight', new_weight=None):
-    for u, v, w in G.edges(data=True):
-        w[new_weight] = w[weight] + dist[u] - dist[v]
+    if G.is_multigraph():
+        get_weight = lambda dist, u, v, data: min(eattr.get(weight, 1)
```
Remove the `dist` parameter. You do not need it. Trust me.

ysitu commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -1044,26 +1077,27 @@ def johnson(G, weight='weight', new_weight=None):
+    for n in G:
+        paths = {n: [n]}  # dictionary of paths
+        all_pairs_path[n] = _dijkstra(G, n, functools.partial(
+            get_weight, dist_bellman), paths=paths)[1]
```
There is no need for `partial` here if you do not use `dist` above.

theosotr commented:
Ok. Done! I hope it is fine now.

ysitu commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -220,42 +219,13 @@ def single_source_dijkstra_path_length(G, source, cutoff=None,
-    return dist
+    if G.is_multigraph():
+        get_weight = lambda u, v, data: min(eattr.get(weight, 1)
+            for eattr in data.values())
```
Wrap the line immediately after `min(` instead of here. Please also review other instances of `get_weight`.

ysitu commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -1044,26 +1076,28 @@ def johnson(G, weight='weight', new_weight=None):
+    if G.is_multigraph():
+        get_weight = lambda u, v, data: min(eattr.get(weight, 1)
+                        for eattr in data.values()) + dist_bellman[u]\
+                                        - dist_bellman[v]
```
This is more readable:
```python
get_weight = lambda u, v, data: (
    min(eattr.get(weight, 1) for eattr in data.values()) +
    dist_bellman[u] - dist_bellman[v])
```

jfinkels commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -1044,26 +1076,28 @@ def johnson(G, weight='weight', new_weight=None):
+    all_pairs_path = {}
+    for n in G:
+        paths = {n: [n]}  # dictionary of paths
+        all_pairs_path[n] = _dijkstra(G, n, get_weight, paths=paths)[1]
```
Maybe simpler as a dictionary comprehension; then it's clear that it can be trivially parallelized (when the time comes to do that): `all_pairs = {v : _dijkstra(G, v, get_weight, paths={n: [n]})[1] for v in G}`.

jfinkels commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -1044,26 +1076,28 @@ def johnson(G, weight='weight', new_weight=None):
+    pred = {}
+    for node in G:
+        dist[node] = 0
+        pred[node] = None
```
Simpler as a dictionary comprehension: `dist = {v: 0 for v in G}` and `pred = {v: None for v in G}`.

theosotr commented:
Ok! Finished.

SanketDG commented on networkx/algorithms/shortest_paths/weighted.py:
```diff
@@ -314,10 +284,66 @@ def single_source_dijkstra(G, source, target=None, cutoff=None, weight='weight')
+    if G.is_directed():
+        G_succ = G.succ
+    else:
+        G_succ = G.adj
```
Just
```python
G_succ = G.succ if G.is_directed() else G.adj
```
